### PR TITLE
Block explorer backend MVP

### PR DESCRIPTION
**Pull request by neysofu**
## Description
This PR adds a new binary crate `block-explorer-backend`, which is an MVP based on Postgres, `axum`, and `sqlx`. The API is inspired from:
- https://github.com/quantstamp/l2-block-explorer-api#1-the-standard
- https://github.com/blockscout/blockscout
- https://jsonapi.org/

You can see at a glance some API patterns in `tests/mod.rs`.

I've tried to isolated changes as much as possible, so that this PR is easy to review as it doesn't modify anything in the rest of the workspace.

### CODE REVIEW

**Comment by neysofu** [comment]
Commit https://github.com/Sovereign-Labs/sovereign-sdk/pull/935/commits/b4563800196f710599fa565ac608b73529cd755b is still not upstreamed but it's technically not part of this PR, I'll rebase once it's upstreamed.

File: block-explorer/backend/src/api_v0/mod.rs [hunk 1]
```diff
@@ -0,0 +1,506 @@
+mod extractors;
```
Comment (bkolad) [inline block-explorer/backend/src/api_v0/mod.rs:1]: Let's add `#![deny(missing_docs)]`

File: block-explorer/backend/src/api_v0/mod.rs [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+        Ok(txs) => txs,
+        Err(err) => {
+            return gateway_timeout_response_504(err);
+        }
+    };
+
+    let full_url = format!("{}{}", state.base_url, uri);
+    let mut links = Links::new(state.base_url.clone());
+    links.add("self", uri.to_string());
+    links.add_pagination_links(&params.pagination.links(&full_url, "FIXME"));
```
Comment (bkolad) [inline block-explorer/backend/src/api_v0/mod.rs]: The above 4 line are repeated in `get_transactions , get_blocks and get_events, get_batches ` should the methods somehow share it?

File: block-explorer/backend/src/indexer.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        .map(|tx| serde_json::to_value(tx).expect("tx is not serializable, this is a bug"))
+        .collect::<Vec<_>>();
+    let batch_range = block.batch_range.clone();
+    let batches = app_state
+        .rpc()
+        .get_batches_range(batch_range.start, batch_range.end, QueryMode::Standard)
+        .await?
+        .into_iter()
+        .map(|batch_opt| serde_json::to_value(batch_opt.expect("No batch")).unwrap())
+        .collect::<Vec<_>>();
```
Comment (bkolad) [inline block-explorer/backend/src/indexer.rs]: It looks like we can avoid collecting here. ` batches` are used inside `upsert_batches`  by `QueryBuildef::push_values` that takes iterator as an argument.  It should be the case also for `block_json` and other variables in this function.

File: block-explorer/backend/src/tests/mod.rs [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+//        .await
+//        .assert_status_ok();
+//    get_transactions_req(&server, &[("page[selection]", "last")])
+//        .await
+//        .assert_status_ok();
+//}
+
+#[tokio::test]
+#[ignore]
+async fn events_first_page_is_full() {
```
Comment (bkolad) [inline block-explorer/backend/src/tests/mod.rs]: Is it possible to move the tests to integration tests in `backend/tests`?

File: block-explorer/backend/src/tests/mod.rs [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+use crate::db::Db;
+use crate::indexer::index_blocks;
+use crate::AppStateInner;
+
+type PostgresContainer<'a> = Container<'a, PostgresImage>;
+
+fn populate_ledger_db() -> LedgerDB {
+    std::env::set_var("TXNS_PER_BLOCK", "10");
+
+    let start_height: u64 = 0u64;
```
Comment (bkolad) [inline block-explorer/backend/src/tests/mod.rs]: Is this code adapted from a benchmark? If yes then maybe it is possible to extract it and use it here and in benchmarks?

File: block-explorer/backend/src/tests/mod.rs (hunk 4 above)
Comment (neysofu) [reply block-explorer/backend/src/tests/mod.rs]: Yes it is, by adding a `lib.rs` and making the crate available as a library. It's been on my TODO list for a while, I'll try to move things around!

File: block-explorer/backend/src/api_v0/mod.rs (hunk 2 above)
Comment (neysofu) [reply block-explorer/backend/src/api_v0/mod.rs]: Indeed! Fixed.

File: block-explorer/backend/src/api_v0/mod.rs (hunk 1 above)
Comment (neysofu) [reply block-explorer/backend/src/api_v0/mod.rs:1]: Done

File: block-explorer/backend/src/tests/mod.rs (hunk 4 above)
Comment (neysofu) [reply block-explorer/backend/src/tests/mod.rs]: Done!

File: block-explorer/backend/src/tests/mod.rs (hunk 5 above)
Comment (neysofu) [reply block-explorer/backend/src/tests/mod.rs]: Yes, it's adapted from `examples/demo-rollup/benches/node`. I've deduplicated some code with PR https://github.com/Sovereign-Labs/sovereign-sdk/pull/1042 but there's some missing. I'll create a tracking issue if it's okay.

Comment (neysofu) [reply block-explorer/backend/src/tests/mod.rs]: Tracked in https://github.com/Sovereign-Labs/sovereign-sdk/issues/1065.

File: block-explorer/backend/src/indexer.rs (hunk 3 above)
Comment (neysofu) [reply block-explorer/backend/src/indexer.rs]: Done! I had to leave `Vec`s and `collect`s in some places in `indexer.rs` but most calls all APIs are not iterator-based.

File: block-explorer/backend/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+clap = { workspace = true, features = ["derive", "env"] }
+hex = "0.4"
+indoc = "2"
+jsonrpsee = { version = "0.20", features = ["client", "client-ws-transport-native-tls"] }
+prometheus = "0.13"
+serde = { version = "1", features = ["derive"] }
+serde_json = { version = "1" }
+serde_urlencoded = "0.7"
+serde_with = "3"
+sov-celestia-adapter = { path = "../../adapters/celestia", features = ["native"], version = "0.2" }
```
Comment (bkolad) [inline block-explorer/backend/Cargo.toml:30]: It would be good to make the code independent on a particular DA. Can we make it genric over `DaSpec`?

Comment (neysofu) [reply block-explorer/backend/Cargo.toml:30]: Done!

**Pull request closed without merging**